File: gsh/gsh_kreatao/report/monthly_roster/monthly_roster.js

frappe.query_reports["Monthly Roster"] = {

    filters: [

        {
            fieldname: "filter_based_on",
            label: __("Filter Based On"),
            fieldtype: "Select",
            options: ["Month", "Date Range"],
            default: "Month",
            reqd: 1,
            on_change: function(report) {
                toggle_filters();
                auto_refresh(report);
            }
        },

        {
            fieldname: "year",
            label: __("Year"),
            fieldtype: "Select",
            options: ["2025","2026","2027","2028","2029","2030","2031","2032","2033","2034","2035"],
            default: (new Date()).getFullYear().toString(),
            depends_on: "eval:doc.filter_based_on == 'Month'",
            on_change: function(report) {
                set_month_date_range();
                auto_refresh(report);
            }
        },

        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: [
                "January","February","March","April","May","June",
                "July","August","September","October","November","December"
            ],
            default: (function() {
                const monthNames = [
                    "January","February","March","April","May","June",
                    "July","August","September","October","November","December"
                ];
                return monthNames[new Date().getMonth()];
            })(),
            depends_on: "eval:doc.filter_based_on == 'Month'",
            on_change: function(report) {
                set_month_date_range();
                auto_refresh(report);
            }
        },

        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            depends_on: "eval:doc.filter_based_on == 'Date Range'",
            on_change: function(report) {
                auto_refresh(report);
            }
        },

        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            depends_on: "eval:doc.filter_based_on == 'Date Range'",
            on_change: function(report) {
                auto_refresh(report);
            }
        },

        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            reqd: 1,
            default: frappe.defaults.get_user_default("Company"),
            on_change: function(report) {
                auto_refresh(report);
            }
        },

        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department",
            on_change: function(report) {
                auto_refresh(report);
            }
        }
    ],

    onload: function(report) {
        toggle_filters();
        set_month_date_range();
        report.refresh();
    }
};

// ===============================
// Toggle Required Fields
// ===============================
function toggle_filters() {

    let mode = frappe.query_report.get_filter_value("filter_based_on");

    let year = frappe.query_report.get_filter("year");
    let month = frappe.query_report.get_filter("month");
    let from_date = frappe.query_report.get_filter("from_date");
    let to_date = frappe.query_report.get_filter("to_date");

    if (mode === "Month") {
        year.df.reqd = 1;
        month.df.reqd = 1;
        from_date.df.reqd = 0;
        to_date.df.reqd = 0;
    } else {
        year.df.reqd = 0;
        month.df.reqd = 0;
        from_date.df.reqd = 1;
        to_date.df.reqd = 1;
    }

    year.refresh();
    month.refresh();
    from_date.refresh();
    to_date.refresh();
}


// ===============================
// Convert Month Name to Date
// ===============================
function set_month_date_range() {

    let mode = frappe.query_report.get_filter_value("filter_based_on");
    if (mode !== "Month") return;

    let year = frappe.query_report.get_filter_value("year");
    let month_name = frappe.query_report.get_filter_value("month");

    if (!year || !month_name) return;

    const monthNames = [
        "January","February","March","April","May","June",
        "July","August","September","October","November","December"
    ];

    let month_index = monthNames.indexOf(month_name);

    if (month_index === -1) return;

    let start_date = new Date(year, month_index, 1);
    let end_date = new Date(year, month_index + 1, 0);

    frappe.query_report.set_filter_value("from_date",
        frappe.datetime.obj_to_str(start_date));

    frappe.query_report.set_filter_value("to_date",
        frappe.datetime.obj_to_str(end_date));
}

//Auto refresh
function auto_refresh(report) {
    clearTimeout(report._refresh_timeout);

    report._refresh_timeout = setTimeout(() => {
        report.refresh();
    }, 300);  // small delay prevents multiple reloads
}

File: gsh/gsh_kreatao/report/monthly_roster/monthly_roster.py

import frappe
from datetime import datetime, timedelta
import calendar

def execute(filters=None):
    if not filters:
        filters = {}

    # Determine date range based on from_date/to_date or month/year
    if filters.get("from_date") and filters.get("to_date"):
        first_day = datetime.strptime(filters["from_date"], "%Y-%m-%d").date()
        last_day = datetime.strptime(filters["to_date"], "%Y-%m-%d").date()
    else:
        year = int(filters.get("year"))
        month_name = filters.get("month")
        month = get_month_number(month_name)
        first_day = datetime(year, month, 1).date()
        last_day = datetime(year, month, calendar.monthrange(year, month)[1]).date()

    columns = get_columns(first_day, last_day)
    data = get_data(filters, first_day, last_day)

    return columns, data

def get_month_number(month_name):
    """Convert month name to month number"""
    return list(calendar.month_name).index(month_name)

def get_columns(first_day, last_day):
    """Generate dynamic columns for the roster based on date range"""
    columns = [{
        "label": "Employee",
        "fieldname": "employee",
        "fieldtype": "Data",
        "width": 180
    }]

    total_days = (last_day - first_day).days + 1
    for i in range(total_days):
        date_obj = first_day + timedelta(days=i)
        day_name = date_obj.strftime("%a")
        label = f"{day_name} {date_obj.day:02d}"
        columns.append({
            "label": label,
            "fieldname": f"day_{i+1}",
            "fieldtype": "HTML",
            "width": 160
        })

    return columns

def get_data(filters, first_day, last_day):
    """Fetch roster data including shifts, leaves, and holidays"""
    company = filters.get("company")
    department = filters.get("department")

    # Employee filters
    employee_conditions = {}
    if department:
        employee_conditions["department"] = department
    if company:
        employee_conditions["company"] = company

    employee_list = frappe.get_all(
        "Employee",
        filters=employee_conditions,
        fields=["name", "employee_name"]
    )

    if not employee_list:
        return []

    employee_ids = [e.name for e in employee_list]

    # Fetch Shift Assignments
    assignments = frappe.db.sql("""
        SELECT
            sa.employee,
            e.employee_name,
            sa.shift_type,
            sa.start_date,
            sa.end_date,
            st.custom_break_start_time,
            st.custom_break_end_time
        FROM `tabShift Assignment` sa
        JOIN `tabEmployee` e ON sa.employee = e.name
        LEFT JOIN `tabShift Type` st ON sa.shift_type = st.name
        WHERE sa.docstatus = 1
        AND sa.employee IN %(employee_ids)s
        AND sa.start_date <= %(last_day)s
        AND (sa.end_date IS NULL OR sa.end_date >= %(first_day)s)
    """, {
        "employee_ids": tuple(employee_ids),
        "first_day": first_day,
        "last_day": last_day
    }, as_dict=1)

    # Fetch Approved Leave
    leaves = frappe.db.sql("""
        SELECT employee, from_date, to_date
        FROM `tabLeave Application`
        WHERE docstatus = 1
        AND status = 'Approved'
        AND employee IN %(employee_ids)s
        AND from_date <= %(last_day)s
        AND to_date >= %(first_day)s
    """, {
        "employee_ids": tuple(employee_ids),
        "first_day": first_day,
        "last_day": last_day
    }, as_dict=1)

    # Fetch Holidays
    holiday_list = frappe.db.get_value("Company", company, "default_holiday_list")
    holidays = frappe.db.sql("""
        SELECT holiday_date, description, weekly_off
        FROM `tabHoliday`
        WHERE parent = %s
        AND holiday_date BETWEEN %s AND %s
    """, (holiday_list, first_day, last_day), as_dict=1)
    holiday_map = {h.holiday_date: h for h in holidays}

    total_days = (last_day - first_day).days + 1
    employee_map = {}

    # Create Leave Map (Optimized)
    leave_map = {}
    for leave in leaves:
        leave_map.setdefault(leave.employee, []).append(leave)

    # Initialize all employees
    for emp in employee_list:
        emp_display = f"{emp.employee_name} ({emp.name})"
        employee_map[emp_display] = {"employee": emp_display}

        for i in range(total_days):
            date_obj = first_day + timedelta(days=i)
            fieldname = f"day_{i+1}"

            # Holiday / Weekly Off
            if date_obj in holiday_map:
                if holiday_map[date_obj].weekly_off:
                    employee_map[emp_display][fieldname] = "<span style='color:blue; font-weight:bold;'>WO</span>"
                else:
                    employee_map[emp_display][fieldname] = f"<span style='color:purple;'>{holiday_map[date_obj].description}</span>"

            # Leave
            elif any(
                    leave.from_date <= date_obj <= leave.to_date
                    for leave in leave_map.get(emp.name, [])
                ):
                
                employee_map[emp_display][fieldname] = "<span style='color:green; font-weight:bold;'>L</span>"

            # Default NA
            else:
                employee_map[emp_display][fieldname] = "<span style='color:red; font-weight:bold;'>NA</span>"

    # Apply Shift Assignments (override only NA)
    for row in assignments:
        emp_display = f"{row.employee_name} ({row.employee})"
        current_date = max(row.start_date, first_day)
        end_date = row.end_date if row.end_date else last_day

        while current_date <= min(end_date, last_day):
            day_index = (current_date - first_day).days + 1
            fieldname = f"day_{day_index}"

            if "NA" in employee_map[emp_display][fieldname]:
                employee_map[emp_display][fieldname] = row.shift_type

            current_date += timedelta(days=1)

    return list(employee_map.values())